syntax = "proto3";

package smartspeech.recognition.v2;

import "google/protobuf/duration.proto";

option go_package = "./;protocol";
option java_package = "TODO";

service SmartSpeech {
  rpc Recognize (stream RecognitionRequest) returns (stream RecognitionResponse);
}

message RecognitionRequest {
  oneof request {
    RecognitionOptions options = 1;
    bytes audio_chunk = 2;
  }
}

enum EouReason {
  UNSPECIFIED = 0;
  ORGANIC = 1;
  NO_SPEECH_TIMEOUT = 2;
  MAX_SPEECH_TIMEOUT = 3;
}

message RecognitionResponse {
  oneof response {
    Transcription transcription = 1;
    BackendInfo backend_info = 2;
    InsightResult insight = 3;
    VADResult vad = 4;
  }
}

message Transcription {
  int32 channel = 1;
  repeated Hypothesis results = 2;
  bool eou = 3;
  EouReason eou_reason = 4;
  google.protobuf.Duration processed_audio_start = 5;
  google.protobuf.Duration processed_audio_end = 6;
  Emotions emotions_result = 7;
  SpeakerInfo speaker_info = 8;
  PersonIdentity person_identity = 9;
}

message InsightResult { string insight_result = 10; }

message VADResult {
  int32 channel = 1;
  google.protobuf.Duration processed_audio_time = 2;
  google.protobuf.Duration utterance_detection_time = 3;
}

message OptionalBool { bool enable = 1; }

message RecognitionOptions {
  enum AudioEncoding {
    AUDIO_ENCODING_UNSPECIFIED = 0;
    PCM_S16LE = 1;
    OPUS = 2;
    MP3 = 3;
    FLAC = 4;
    ALAW = 5;
    MULAW = 6;
    G729 = 7;
  }

  AudioEncoding audio_encoding = 1;
  int32 sample_rate = 2;
  int32 channels_count = 3;
  string language = 4;
  string model = 5;
  OptionalBool enable_multi_utterance = 6;
  OptionalBool enable_partial_results = 7;
  int32 hypotheses_count = 8;
  google.protobuf.Duration no_speech_timeout = 9;
  google.protobuf.Duration max_speech_timeout = 10;
  Hints hints = 11;
  SpeakerSeparationOptions speaker_separation_options = 12;
  NormalizationOptions normalization_options = 13;
  repeated string insight_models = 14;
  OptionalBool enable_vad = 15;
  OptionalBool custom_ws_flow_control = 16;
  OptionalBool enable_long_utterances = 17;
  repeated string truncate_last_punctuation = 18;
}

message NormalizationOptions {
  OptionalBool enable = 1;
  OptionalBool profanity_filter = 2;
  OptionalBool punctuation = 3;
  OptionalBool capitalization = 4;
  OptionalBool question = 5;
  OptionalBool force_cyrillic = 6;
}

message Hints {
  repeated string words = 1;
  bool enable_letters = 2;
  google.protobuf.Duration eou_timeout = 3;
}

message SpeakerSeparationOptions {
  bool enable = 1;
  bool enable_only_main_speaker = 2;
  int32 count = 3;
}

message Hypothesis {
  message WordAlignment {
    string word = 1;
    google.protobuf.Duration start = 2;
    google.protobuf.Duration end = 3;
  }

  string text = 1;
  string normalized_text = 2;
  google.protobuf.Duration start = 3;
  google.protobuf.Duration end = 4;
  repeated WordAlignment word_alignments = 5;
}

message Emotions {
  float positive = 1;
  float neutral = 2;
  float negative = 3;
  float positive_a = 4;
  float neutral_a = 5;
  float negative_a = 6;
  float positive_t = 7;
  float neutral_t = 8;
  float negative_t = 9;
}

message BackendInfo {
  string model_name = 1;
  string model_version = 2;
  string server_version = 3;
}

message SpeakerInfo {
  int32 speaker_id = 1;
  float main_speaker_confidence = 2;
}

enum AgeType { AGE_NONE = 0; CHILD = 1; ADULT = 2; }
enum GenderType { GENDER_NONE = 0; MALE = 1; FEMALE = 2; }

message PersonIdentity {
  AgeType age = 1;
  GenderType gender = 2;
  float age_score = 3;
  float gender_score = 4;
}


